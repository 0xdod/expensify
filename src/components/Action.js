import React from "react";

function Action(props) {
  return (
    <div>
      <button className="big-button" onClick={props.handleRandomPick} disabled={!props.hasOptions}>
        What do you want to do today?
      </button>
    </div>
  );
}

export default Action;
