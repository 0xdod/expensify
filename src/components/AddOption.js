import React, { Component } from 'react';

class AddOption extends Component {
  state = {
    error: null,
  };

  handleAddOption = e => {
    e.preventDefault();
    const optionNode = e.target.elements.option;
    const option = optionNode.value.trim();

    const error = this.props.handleAddOption(option);
    this.setState(() => ({ error }))
    optionNode.value = '';
  };

  render() {
    return (
      <div>
        {this.state.error && (
          <p className="add-option-error">
          {this.state.error}</p>
        )}
        <form className="add-option" onSubmit={this.handleAddOption}>
          <input className="add-option__input" type="text" name="option" />
          <button className="button" type="submit">Add option</button>
        </form>
      </div>
    );
  }
}

export default AddOption;
