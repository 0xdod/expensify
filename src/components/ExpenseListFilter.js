import React from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate,
} from '../actions/filters';

export class ExpenseListFilter extends React.Component {
	state = {
		calendarFocused: null,
	};

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	onFocusChange = calendarFocused => {
		this.setState(() => ({ calendarFocused }));
	};
	onTextChange = e => {
		this.props.setTextFilter(e.target.value);
	};
	onSortChange = e => {
		if (e.target.value === 'date') {
			this.props.sortByDate();
		} else if (e.target.value === 'amount') {
			this.props.sortByAmount();
		} else return false;
	};

	render() {
		return (
			<div>
				<input
					type="text"
					value={this.props.filters.text}
					onChange={this.onTextChange}
				/>
				<select
					value={this.props.filters.sortBy}
					onChange={this.onSortChange}
				>
					{/*<option value="">---</option>*/}
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonth={1}
					isOutsideRange={() => false}
					showClearDates={true}
				/>
			</div>
		);
	}
}
const mapStateToProps = state => ({
	filters: state.filters,
});

const mapDispatchToProps = dispatch => ({
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: date => dispatch(setStartDate(date)),
	setEndDate: date => dispatch(setEndDate(date)),
	setTextFilter: value => dispatch(setTextFilter(value)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
