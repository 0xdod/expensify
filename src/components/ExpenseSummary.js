import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral'

import selectExpenseTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

function pluralize(len) {
	if (len > 1)
		return 's'
	return ''
}

export const ExpenseSummary = ({expenseCount, expenseTotal}) => (
	<div>
        <p>
           Viewing {expenseCount} {'expense'+pluralize(expenseCount)} totalling {numeral(expenseTotal / 100).format('$0,0.00')}
        </p>
	</div>
);

const mapStateToProps = (state) => {
	const visibleExpenses = selectExpenses(state.expenses, state.filters)
	return {
		expenseTotal: selectExpenseTotal(visibleExpenses),
		expenseCount: visibleExpenses.length, 
	}
}

export default connect(mapStateToProps)(ExpenseSummary);