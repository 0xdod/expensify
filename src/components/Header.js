import React from 'react';
import { NavLink } from 'react-router-dom';

const Seperator = () => (<span>&nbsp;|&nbsp;</span>)

const Header = () => (
	<header>
		<h1>Expensify</h1>
		<NavLink to="/" exact={true} activeClassName="is-active">
			Dashboard
		</NavLink>
		<Seperator />
		<NavLink to="/create" activeClassName="is-active">
			Create
		</NavLink>
		<Seperator />
		<NavLink to="/help" activeClassName="is-active">
			Help
		</NavLink>
	</header>
);

export default Header;
