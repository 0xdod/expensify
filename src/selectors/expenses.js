import moment from 'moment'

export default (expenses, filters) => {
	const { text = '', sortBy = 'date', startDate = undefined, endDate = undefined } = filters;
	return expenses
		.filter(expense => {
			const createdAtMoment = moment(expense.createdAt)
			const startDateMatch = startDate ? 
			startDate.isSameOrBefore(createdAtMoment, 'day') : true;
			const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day') : true;
			const descMatch = expense.description
				.toLowerCase()
				.includes(text.toLowerCase());
			const noteMatch = expense.note
				.toLowerCase()
				.includes(text.toLowerCase());
			let textMatch = descMatch || noteMatch;
			return startDateMatch && endDateMatch && textMatch;
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1;
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1;
			}
			return 0
		});
};
