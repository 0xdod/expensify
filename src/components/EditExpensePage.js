import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { removeExpense, editExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
	onSubmit = expense => {
		this.props.editExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};

	onRemove = id => {
		this.props.removeExpense(id);
		this.props.history.push('/');
	};
	render() {
		return (
			<div>
				<ExpenseForm
					expense={this.props.expense}
					onSubmit={expense => this.onSubmit(expense)}
				/>
				<button onClick={ () => this.onRemove(this.props.expense.id)}>
					Remove
				</button>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find(
			expense => expense.id == props.match.params.id
		),
	};
};

const mapDispatchToProps = dispatch => {
	return {
		editExpense: (id, expense) => dispatch(editExpense(id, expense)),
		removeExpense: id => dispatch(removeExpense(id)),
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
