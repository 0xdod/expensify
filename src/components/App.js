import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Header from './Header';
import ExpenseDashboard from './ExpenseDashboardPage';
import AddExpensePage from './AddExpensePage';
import EditExpensePage from './EditExpensePage';
import HelpPage from './HelpPage';
import NotFoundPage from './NotFoundPage';

const App = () => (
	<Router>
		<div>
			<Header />
			<Switch>
				<Route path="/" component={ExpenseDashboard} exact={true} />
				<Route path="/create" component={AddExpensePage} />
				<Route path="/edit/:id" component={EditExpensePage} />
				<Route path="/help" component={HelpPage} />
				<Route component={NotFoundPage} />
			</Switch>
		</div>
	</Router>
);

export default App;
