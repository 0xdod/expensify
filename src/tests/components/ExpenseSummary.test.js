import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary'
import expenses from '../fixtures/expenses';

test('should render ExpenseSummary correctly with multiple count', () => {
	const wrapper = shallow(<ExpenseSummary count={5} total={1000} />)
	expect(toJSON(wrapper)).toMatchSnapshot()
})

test('should render ExpenseSummary correctly with a single count', () => {
	const wrapper = shallow(<ExpenseSummary count={1} total={100} />)
	expect(toJSON(wrapper)).toMatchSnapshot()
})