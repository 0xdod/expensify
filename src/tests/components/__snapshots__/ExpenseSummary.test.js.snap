// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with a single count 1`] = `
<div>
  <p>
    Viewing 
     
    expense
     totalling 
    $0.00
  </p>
</div>
`;

exports[`should render ExpenseSummary correctly with multiple count 1`] = `
<div>
  <p>
    Viewing 
     
    expense
     totalling 
    $0.00
  </p>
</div>
`;
